File: go/center_quantiles_impl.go
package pragmastat

import (
	"math"
	"sort"
)

// Every multiply-add in this file is written as float64(a*b) + float64(c*d)
// rather than a*b + c*d. The conversions are not redundant. The Go
// specification lets an implementation fuse a multiply into an add as a single
// rounding, and gc does it on arm64, ppc64le, s390x, riscv64 and loong64,
// never on amd64. None of the other six languages fuses, so a fused result
// would differ in the last bit from every one of them. An explicit conversion
// is the only way the language offers to pin the intermediate rounding.
//
// Here the multiplier is always 0.5, so the scaling is exact and the fused and
// unfused forms agree bit for bit: none of these sites was ever wrong. They are
// pinned for uniformity, so this file needs no entry on the go:check:fma
// exemption list and no future reader has to re-derive that proof to know the
// file is safe.

// relativeEpsilon is the tolerance for floating-point comparisons in binary search convergence.
const relativeEpsilon = 1e-14

// centerQuantileBoundsImpl computes both lower and upper bounds from pairwise averages.
// Uses binary search with counting function to avoid materializing all N(N+1)/2 pairs.
func centerQuantileBoundsImpl(sorted []float64, marginLo, marginHi int64) (lo, hi float64) {
	n := len(sorted)
	totalPairs := int64(n) * int64(n+1) / 2

	if marginLo < 1 {
		marginLo = 1
	}
	if marginLo > totalPairs {
		marginLo = totalPairs
	}
	if marginHi < 1 {
		marginHi = 1
	}
	if marginHi > totalPairs {
		marginHi = totalPairs
	}

	lo = centerFindExactQuantileImpl(sorted, marginLo)
	hi = centerFindExactQuantileImpl(sorted, marginHi)

	if lo > hi {
		lo, hi = hi, lo
	}
	return lo, hi
}

// centerCountPairsLessOrEqualImpl counts pairwise averages <= target value.
// Uses O(n) two-pointer algorithm.
func centerCountPairsLessOrEqualImpl(sorted []float64, target float64) int64 {
	n := len(sorted)
	var count int64
	// j is not reset: as i increases, threshold decreases monotonically
	j := n - 1

	for i := range n {
		threshold := 2*target - sorted[i]

		for j >= 0 && sorted[j] > threshold {
			j--
		}

		if j >= i {
			count += int64(j - i + 1)
		}
	}

	return count
}

// centerFindExactQuantileImpl finds the exact k-th pairwise average using selection algorithm.
func centerFindExactQuantileImpl(sorted []float64, k int64) float64 {
	n := len(sorted)
	totalPairs := int64(n) * int64(n+1) / 2

	if n == 1 {
		return sorted[0]
	}

	if k == 1 {
		return sorted[0]
	}

	if k == totalPairs {
		return sorted[n-1]
	}

	lo := sorted[0]
	hi := sorted[n-1]
	const eps = relativeEpsilon

	for hi-lo > eps*math.Max(1.0, math.Max(math.Abs(lo), math.Abs(hi))) {
		// Overflow-safe, order-symmetric midpoint: 0.5*a + 0.5*b (halve before
		// summing; never overflows; operand order is irrelevant).
		mid := float64(0.5*lo) + float64(0.5*hi)
		countLessOrEqual := centerCountPairsLessOrEqualImpl(sorted, mid)

		if countLessOrEqual >= k {
			hi = mid
		} else {
			lo = mid
		}
	}

	// Overflow-safe, order-symmetric midpoint: 0.5*a + 0.5*b (halve before
	// summing; never overflows; operand order is irrelevant).
	target := float64(0.5*lo) + float64(0.5*hi)
	var candidates []float64

	for i := range n {
		threshold := 2*target - sorted[i]

		left := i
		right := n

		for left < right {
			m := (left + right) / 2
			if sorted[m] < threshold-eps {
				left = m + 1
			} else {
				right = m
			}
		}

		if left < n && left >= i && math.Abs(sorted[left]-threshold) < eps*math.Max(1.0, math.Abs(threshold)) {
			candidates = append(candidates, float64(0.5*sorted[i])+float64(0.5*sorted[left]))
		}

		if left > i {
			avgBefore := float64(0.5*sorted[i]) + float64(0.5*sorted[left-1])
			if avgBefore <= target+eps {
				candidates = append(candidates, avgBefore)
			}
		}
	}

	if len(candidates) == 0 {
		return target
	}

	// Sort candidates
	sort.Float64s(candidates)

	for _, candidate := range candidates {
		countAtCandidate := centerCountPairsLessOrEqualImpl(sorted, candidate)
		if countAtCandidate >= k {
			return candidate
		}
	}

	return target
}

File: r/pragmastat/R/center_quantiles_impl.R
# Algorithm for computing specific quantiles of pairwise averages.
# Uses binary search with counting to avoid materializing all N(N+1)/2 pairs.

# Relative tolerance for convergence
.RELATIVE_EPSILON <- 1e-14

# Computes both lower and upper bounds from pairwise averages.
#
# @param sorted Sorted numeric vector
# @param margin_lo 1-based rank for lower bound
# @param margin_hi 1-based rank for upper bound
# @return List with 'lower' and 'upper' components
center_quantile_bounds_impl <- function(sorted, margin_lo, margin_hi) {
  n <- length(sorted)
  total_pairs <- as.numeric(n) * as.numeric(n + 1) / 2

  if (margin_lo < 1) margin_lo <- 1
  if (margin_lo > total_pairs) margin_lo <- total_pairs
  if (margin_hi < 1) margin_hi <- 1
  if (margin_hi > total_pairs) margin_hi <- total_pairs

  lo <- center_find_exact_quantile_impl(sorted, margin_lo)
  hi <- center_find_exact_quantile_impl(sorted, margin_hi)

  if (lo > hi) {
    tmp <- lo
    lo <- hi
    hi <- tmp
  }

  return(list(lower = lo, upper = hi))
}

# Counts pairwise averages <= target value using O(n) two-pointer algorithm.
center_count_pairs_impl <- function(sorted, target) {
  n <- length(sorted)
  count <- 0
  # j is not reset: as i increases, threshold decreases monotonically
  j <- n

  for (i in 1:n) {
    threshold <- 2 * target - sorted[i]

    while (j >= 1 && sorted[j] > threshold) {
      j <- j - 1
    }

    if (j >= i) {
      count <- count + (j - i + 1)
    }
  }

  return(count)
}

# Finds the exact k-th pairwise average using binary search + candidate refinement.
center_find_exact_quantile_impl <- function(sorted, k) {
  n <- length(sorted)
  total_pairs <- as.numeric(n) * as.numeric(n + 1) / 2

  if (n == 1) {
    return(sorted[1])
  }
  if (k == 1) {
    return(sorted[1])
  }
  if (k == total_pairs) {
    return(sorted[n])
  }

  lo <- sorted[1]
  hi <- sorted[n]
  eps <- .RELATIVE_EPSILON

  while (hi - lo > eps * max(1.0, abs(lo), abs(hi))) {
    mid <- 0.5 * lo + 0.5 * hi
    count_le <- center_count_pairs_impl(sorted, mid)

    if (count_le >= k) {
      hi <- mid
    } else {
      lo <- mid
    }
  }

  target <- 0.5 * lo + 0.5 * hi
  candidates <- numeric(0)

  for (i in 1:n) {
    threshold <- 2 * target - sorted[i]

    # Binary search for insertion point
    left <- i
    right <- n + 1

    while (left < right) {
      m <- (left + right) %/% 2
      if (sorted[m] < threshold - eps) {
        left <- m + 1
      } else {
        right <- m
      }
    }

    if (left <= n && left >= i && abs(sorted[left] - threshold) < eps * max(1.0, abs(threshold))) {
      candidates <- c(candidates, 0.5 * sorted[i] + 0.5 * sorted[left])
    }

    if (left > i) {
      avg_before <- 0.5 * sorted[i] + 0.5 * sorted[left - 1]
      if (avg_before <= target + eps) {
        candidates <- c(candidates, avg_before)
      }
    }
  }

  if (length(candidates) == 0) {
    return(target)
  }

  candidates <- sort(candidates)

  for (candidate in candidates) {
    count_at <- center_count_pairs_impl(sorted, candidate)
    if (count_at >= k) {
      return(candidate)
    }
  }

  return(target)
}
